src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs

```diff
@@ -568,15 +570,61 @@ public virtual void WriteLiteralTo([NotNull] TextWriter writer, string value)
                     // Calculate length of the source span by the position of the next value (or suffix)
                     var sourceLength = next.Position - attributeValue.Value.Position;
 
-                    var stringValue = positionTaggedAttributeValue.Value as string;
-
-                    WriteAttributeValue(writer, attributeValue, stringValue, sourceLength);
+                    WriteAttributeValue(writer, attributeValue, sourceLength);
                 }
             }
 
             WritePositionTaggedLiteral(writer, suffix);
         }
 
+        public void AddHtmlAttributeValues(
+            string attributeName,
+            TagHelperExecutionContext executionContext,
+            params AttributeValue[] values)
+        {
+            bool useAttributeNameAsValue;
+            if (!ShouldRenderAttribute(values, out useAttributeNameAsValue))
+            {
+                // The first value was 'null' or 'false' indicating that we shouldn't render the attribute. The
+                // attribute is treated as a TagHelper attribute so it's only available in
+                // TagHelperContext.AllAttributes for TagHelper authors to see (if they want to see why the attribute
+                // was removed from TagHelperOutput.Attributes).
+                executionContext.AddTagHelperAttribute(
+                    attributeName,
+                    values[0].Value.Value?.ToString() ?? string.Empty);
+
+                return;
+            }
+            else if (useAttributeNameAsValue)
+            {
+                executionContext.AddHtmlAttribute(attributeName, attributeName);
+            }
+            else
+            {
+                var valueBuffer = new StringCollectionTextWriter(Output.Encoding);
+
+                foreach (var value in values)
+                {
+                    if (value.Value.Value == null)
+                    {
+                        // Skip null values
+                        continue;
+                    }
+
+                    if (!string.IsNullOrEmpty(value.Prefix))
+                    {
+                        WriteLiteralTo(valueBuffer, value.Prefix);
+                    }
+
+                    WriteUnprefixedAttributeValueTo(valueBuffer, value);
+                }
+
+                var htmlString = new HtmlString(valueBuffer.ToString());
```

Comment:
Ick. We should avoid smashing the `StringCollectionTextWriter`'s buffer together here. Suggest filing a follow-up bug to fix this e.g. add an `HtmlString` constructor overload that accepts an `IHtmlContext`. May be easier to clean up once @sornaks gets aspnet/Razor#473 and #2899  in.